sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import java.util.{Iterator => JIterator}
+
+import org.apache.spark.sql.catalyst.expressions.{SortOrder, Ascending, JoinedRow}
+import org.apache.spark.sql.catalyst.expressions.codegen.GenerateOrdering
+import org.apache.spark.sql.execution.QueryExecution
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.types.StructType
+
+import org.apache.spark.api.java.function.{Function => JFunction, Function2 => JFunction2, _}
+
+import scala.reflect.runtime.universe._
+
+/**
+ * A typed collection of data stored efficiently as a DataFrame.
+ */
+@Experimental
+class Dataset[T] private[sql](
+    @transient val sqlContext: SQLContext,
+    @transient val queryExecution: QueryExecution)(
+    implicit val encoder: Encoder[T]) extends Serializable {
+
+  /**
+   * Returns a new `Dataset` where each record has been mapped on to the specified type.
+   */
+  def as[U : Encoder] = new Dataset(sqlContext, queryExecution)(implicitly[Encoder[U]])
+
+  /**
+   * Concise syntax for chaining custom transformations.
+   * {{{
+   *   dataset
+   *     .transform(featurize)
+   *     .transform(...)
+   * }}}
+   */
+  def transform[U](t: Dataset[T] => Dataset[U]): Dataset[U] = ???
```

Comment:
Thats it.  It's sugar so you can write:

```
dataset
  .where(...)
  .transform(featurize)
  .transform(tokenize)
```

instead of `featurize(tokenize(ds.where(...))))`